coverage: Allow each coverage statement to have multiple code regions

Zalathar opened the pull request:
The original implementation of coverage instrumentation was built around the assumption that a coverage counter/expression would be associated with *up to one* code region. When it was discovered that *multiple* regions would sometimes need to share a counter, a workaround was found: for the remaining regions, the instrumentor would create a fresh expression that adds zero  to the existing counter/expression.

That got the job done, but resulted in some awkward code, and produces unnecessarily complicated coverage maps in the final binary.

---

This PR removes that tension by changing `StatementKind::Coverage`'s code region field from `Option<CodeRegion>` to `Vec<CodeRegion>`.

The changes on the codegen side are fairly straightforward. As long as each `CoverageKind::Counter` only injects one `llvm.instrprof.increment`, the rest of coverage codegen is happy to handle multiple regions mapped to the same counter/expression, with only minor option-to-vec adjustments.

On the instrumentor/mir-transform side, we can get rid of the code that creates extra (x + 0) expressions. Instead we gather all of the code regions associated with a single BCB, and inject them all into one coverage statement.

---

There are several patches here but they can be divided in to three phases:
- Preparatory work
- Actually switching over to multiple regions per coverage statement
- Cleaning up

So viewing the patches individually may be easier.

Zalathar committed:
coverage: Let each coverage statement hold a vector of code regions

This makes it possible for a `StatementKind::Coverage` to hold more than one
code region, but that capability is not yet used.

Zalathar committed:
coverage: Remove code for making expression copies of BCB counters

Now that coverage statements can have multiple code regions attached to them,
this code is never used.

Zalathar commented:
(Cargo.toml changes are part of #114843 and won't remain in the final version of this PR.)

Zalathar commented on compiler/rustc_codegen_llvm/src/coverageinfo/mod.rs:
```diff
@@ -110,25 +110,15 @@ impl<'tcx> CoverageInfoBuilderMethods<'tcx> for Builder<'_, '_, 'tcx> {
-                    debug!(
-                        "adding counter to coverage_map: instance={:?}, id={:?}, region={:?}",
-                        instance, id, code_region,
-                    );
```
These `debug!` invocations have been replaced with `#[instrument(...)]` tracing on the methods being called.

Zalathar commented on compiler/rustc_codegen_llvm/src/coverageinfo/mod.rs:
```diff
@@ -110,25 +110,15 @@ impl<'tcx> CoverageInfoBuilderMethods<'tcx> for Builder<'_, '_, 'tcx> {
-                    );
-                    func_coverage.add_counter(id, code_region);
-                }
+                func_coverage.add_counter(id, code_regions);
```
We now call `add_counter` unconditionally here, and leave the callee to decide how to handle counters with zero code regions.

Swatinem commented on compiler/rustc_codegen_llvm/src/coverageinfo/map_data.rs:
```diff
@@ -156,156 +223,66 @@ impl<'tcx> FunctionCoverage<'tcx> {
+            .iter_enumerated()
+            // Filter out counter IDs that we never saw during MIR traversal.
+            // This can happen if a counter was optimized out by MIR transforms.
+            // (Its regions will have been remapped to zero instead.)
```
```suggestion
            // (It was replaced by an `Unreachable` mapping)
```

Zalathar commented on tests/coverage-map/status-quo/async.cov-map:
```diff
@@ -148,14 +146,14 @@ Number of file 0 mappings: 1
+- Code(Counter(0)) at (prev + 21, 23) to (start + 1, 12)
 - Code(Zero) at (prev + 2, 9) to (start + 0, 10)
-- Code(Counter(0)) at (prev + 0, 14) to (start + 0, 17)
+- Code(Zero) at (prev + 0, 14) to (start + 0, 17)
```
There's some existing code in `add_unused_function_coverage` that ensures each unused function's coverage has at least one `Counter` mapping.

(I'm not convinced it's necessary, but I don't want to do the homework of proving that it's *not* necessary, so for now I've left it as-is.)

davidtwco commented:
There's still lots of discussion and changes happening here, use `@rustbot ready` when you've got something you're looking to have merged.

bors commented:
:umbrella: The latest upstream changes (presumably #115183) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #115058) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #114843) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Zalathar commented:
@rustbot label -A-testsuite -T-bootstrap

Zalathar committed:
coverage: Remove `next_id` methods from counter/expression IDs

When these methods were originally written, I wasn't aware that
`newtype_index!` already supports addition with ordinary numbers, without
needing to unwrap and re-wrap.

Zalathar committed:
coverage: Mappings for unused functions can all be zero

There is no need to include a dummy counter reference in the coverage mappings
for an unused function.

bors commented:
:umbrella: The latest upstream changes (presumably #115586) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #115671) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #115894) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Zalathar committed:
coverage: Update comments/logs that referred to `CoverageSpan`

The concrete type `CoverageSpan` is no longer used outside of the `spans`
module.

This is a separate patch to avoid noise in the preceding patch that actually
encapsulates coverage spans.

Zalathar committed:
coverage: Encapsulate coverage spans

By encapsulating the coverage spans in a struct, we can change the internal
representation without disturbing existing call sites. This will be useful for
grouping coverage spans by BCB.

This patch includes some changes that were originally in #115912, which avoid
the need for a particular test to deal with coverage spans at all.

(Comments/logs referring to `CoverageSpan` are updated in a subsequent patch.)

Zalathar commented:
I've folded some `spans` changes into this [(diff)](https://github.com/rust-lang/rust/compare/2037b9860bccb4c3b704d994bf5735e43f5b3700..5ede04f1ba6cd2b6c0cc92f455aca6827ef27376), because they make it easier to get all of the coverage spans associated with a BCB.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6221163156/job/16882601810) [(plain)](https://github.com/rust-lang/rust/commit/3eb275090e607352a13e7020f5296a5925bdfbea/checks/16882601810/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=Zalathar
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_452febd8-9e1f-43d5-ac52-1b5ceca35c1e
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=a4fa6cf9213e0f5e2d0eb171e8692bf08ed0f5a6
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_452febd8-9e1f-43d5-ac52-1b5ceca35c1e
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_452febd8-9e1f-43d5-ac52-1b5ceca35c1e
GITHUB_TRIGGERING_ACTOR=Zalathar
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/115301/merge
GITHUB_WORKFLOW_SHA=a4fa6cf9213e0f5e2d0eb171e8692bf08ed0f5a6
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_18_X64=/opt/hostedtoolcache/go/1.18.10/x64
---
    Checking rustc_mir_transform v0.0.0 (/checkout/compiler/rustc_mir_transform)
error[E0308]: mismatched types
  --> compiler/rustc_mir_transform/src/coverage/spans.rs:54:19
   |
54 |             .map(|(bcb, coverage_spans)| (*bcb, coverage_spans.as_slice()))
   |                   |
   |                   |
   |                   expected `Vec<CoverageSpan>`, found `(_, _)`
   |
   |
   = note: expected struct `Vec<CoverageSpan>`
               found tuple `(_, _)`
For more information about this error, try `rustc --explain E0308`.
error: could not compile `rustc_mir_transform` (lib) due to previous error
warning: build failed, waiting for other jobs to finish...
error: could not compile `rustc_mir_transform` (lib test) due to previous error
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #115952) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Zalathar committed:
coverage: Store each BCB's code regions in one coverage statement

If a BCB has more than one code region, those extra regions can now all be
stored in the same coverage statement, instead of being stored in additional
statements.

bors commented:
:umbrella: The latest upstream changes (presumably #115865) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Zalathar commented:
#114399 has landed, so I'm going to give this PR another round of polish and then flag it for review.

Zalathar commented:
@davidtwco All dependencies have landed, so this is now ready for review.
@rustbot ready

Zalathar commented:
(I've been rebasing this for the benefit of other PRs/patches that rely on it.)

davidtwco reviewed (approved):
Apologies for the delay in getting to this, looks good to me.

davidtwco commented:
@bors r+

bors commented:
:pushpin: Commit 053c4f94a098be5fbb80ca881b01cc8e9c64a8a5 has been approved by `davidtwco`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=davidtwco 053c4f94a098be5fbb80ca881b01cc8e9c64a8a5 -->
<!-- homu: {"type":"Approved","sha":"053c4f94a098be5fbb80ca881b01cc8e9c64a8a5","approver":"davidtwco","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 053c4f94a098be5fbb80ca881b01cc8e9c64a8a5 with merge 36aab8df0ab8a76f1c4f95ce8becefdd8a6fe526...
<!-- homu: {"type":"BuildStarted","head_sha":"053c4f94a098be5fbb80ca881b01cc8e9c64a8a5","merge_sha":"36aab8df0ab8a76f1c4f95ce8becefdd8a6fe526"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6397059563/job/17367582120)
Approved by: davidtwco
Pushing 36aab8df0ab8a76f1c4f95ce8becefdd8a6fe526 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"davidtwco","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6397059563/job/17367582120"},"merge_sha":"36aab8df0ab8a76f1c4f95ce8becefdd8a6fe526"} -->

rust-timer commented:
Finished benchmarking commit (36aab8df0ab8a76f1c4f95ce8becefdd8a6fe526): [comparison URL](https://perf.rust-lang.org/compare.html?start=268d6250299b4f08f8e669a2889fdf71388d1f39&end=36aab8df0ab8a76f1c4f95ce8becefdd8a6fe526&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=268d6250299b4f08f8e669a2889fdf71388d1f39&end=36aab8df0ab8a76f1c4f95ce8becefdd8a6fe526&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.7% | [-0.7%, -0.6%] | 4     |
| Improvements ✅ <br /> (secondary) | -0.4% | [-0.6%, -0.2%] | 15    |
| All ❌✅ (primary)                 | -0.7% | [-0.7%, -0.6%] | 4     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=268d6250299b4f08f8e669a2889fdf71388d1f39&end=36aab8df0ab8a76f1c4f95ce8becefdd8a6fe526&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 1.1%  | [1.1%, 1.1%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.8% | [-0.8%, -0.8%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.8% | [-0.8%, -0.8%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=268d6250299b4f08f8e669a2889fdf71388d1f39&end=36aab8df0ab8a76f1c4f95ce8becefdd8a6fe526&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=268d6250299b4f08f8e669a2889fdf71388d1f39&end=36aab8df0ab8a76f1c4f95ce8becefdd8a6fe526&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 622.003s -> 621.563s (-0.07%)
**Artifact size**: 272.02 MiB -> 272.01 MiB (-0.00%)
